src/timely-util/src/containers/array.rs

```diff
@@ -120,6 +128,22 @@ impl<T> Deref for Array<T> {
     }
 }
 
+impl<T> DerefMut for Array<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        // TODO: Use `slice_assume_init_ref` once stable.
+        // Context: https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.slice_assume_init_ref
+        // The following safety argument is adapted from the source.
+        // SAFETY: casting `elements` to a `*const [T]` is safe since the caller guarantees that
+        // `slice` is initialized, and `MaybeUninit` is guaranteed to have the same layout as `T`.
+        // The pointer obtained is valid since it refers to memory owned by `elements` which is a
+        // reference and thus guaranteed to be valid for reads.
```

Comment:
This safety argument talks about `*const` and reads, but we're casting to a `*mut` and deference that for writes. I think a similar argument still holds, based on that `elements` is mutably borrowed here, so we can produce a `&mut` into it.

Reply:
Fortunately, that code is gone :)

Reply:
Fortunately, that code is gone :)